### PR TITLE
Simplify the K8sExecutor and K8sPodOperator

### PR DESCRIPTION
**Pull request by dimberman**
As discussed in an earlier email thread, this PR removes much of the configuration elements of the K8sExecutor for a much simpler design.

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).

### CODE REVIEW

File: airflow/kubernetes/pod_generator.py [hunk 1]
```diff
@@ -96,100 +96,18 @@ class PodGenerator:
...
-        tolerations: Optional[list] = None,
-        security_context: Optional[Union[k8s.V1PodSecurityContext, dict]] = None,
-        configmaps: Optional[List[str]] = None,
-        dnspolicy: Optional[str] = None,
-        schedulername: Optional[str] = None,
         pod: Optional[k8s.V1Pod] = None,
         pod_template_file: Optional[str] = None,
-        extract_xcom: bool = False,
-        priority_class_name: Optional[str] = None,
+        extract_xcom=True
```
Comment (turbaszek) [inline airflow/kubernetes/pod_generator.py]:
```suggestion
        extract_xcom: bool =True
```

File: airflow/providers/cncf/kubernetes/operators/kubernetes_pod.py [hunk 2]
```diff
@@ -175,7 +175,7 @@ def __init__(self,  # pylint: disable=too-many-arguments,too-many-locals
                  affinity: Optional[Dict] = None,
                  config_file: Optional[str] = None,
                  node_selectors: Optional[Dict] = None,
-                 image_pull_secrets: Optional[str] = None,
+                 image_pull_secrets: List[k8s.V1LocalObjectReference] = None,
```
Comment (turbaszek) [inline airflow/providers/cncf/kubernetes/operators/kubernetes_pod.py]:
```suggestion
                 image_pull_secrets: Optional[List[k8s.V1LocalObjectReference]] = None,
```

File: airflow/providers/cncf/kubernetes/operators/kubernetes_pod.py [hunk 3]
```diff
@@ -365,44 +365,57 @@ def create_new_pod_for_operator(self, labels, launcher) -> Tuple[State, k8s.V1Po
...
-            dnspolicy=self.dnspolicy,
-            schedulername=self.schedulername,
-            init_containers=self.init_containers,
-            restart_policy='Never',
-            priority_class_name=self.priority_class_name,
-            pod_template_file=self.pod_template_file,
-            pod=self.full_pod_spec,
-        ).gen_pod()
+        from kubernetes.client import models as k8s
+        pod = None
```
Comment (turbaszek) [inline airflow/providers/cncf/kubernetes/operators/kubernetes_pod.py]:
```suggestion

```
Probably no need for that as we have if/else

File: airflow/providers/cncf/kubernetes/operators/kubernetes_pod.py [hunk 4]
```diff
@@ -365,44 +365,57 @@ def create_new_pod_for_operator(self, labels, launcher) -> Tuple[State, k8s.V1Po
...
-            security_context=self.security_context,
-            dnspolicy=self.dnspolicy,
-            schedulername=self.schedulername,
-            init_containers=self.init_containers,
-            restart_policy='Never',
-            priority_class_name=self.priority_class_name,
-            pod_template_file=self.pod_template_file,
-            pod=self.full_pod_spec,
-        ).gen_pod()
+        from kubernetes.client import models as k8s
```
Comment (turbaszek) [inline airflow/providers/cncf/kubernetes/operators/kubernetes_pod.py]: Any particular reason to import this locally?

**Review by turbaszek — COMMENTED** [review]
-2.5k of lines - love it ❤️ This is still WIP but we should remember about note in UPDATING.md

**Comment by potiuk** [comment]
> -2.5k of lines - love it  This is still WIP but we should remember about note in UPDATING.md

Indeed. I love the stats on this one.

**Comment by kaxil** [comment]
Nice <3 -- some static checks are failing https://github.com/apache/airflow/pull/10393/checks?check_run_id=1003310026

File: airflow/providers/cncf/kubernetes/operators/kubernetes_pod.py (hunk 4 above)
Comment (dimberman) [reply airflow/providers/cncf/kubernetes/operators/kubernetes_pod.py]: Nope removed that

File: tests/kubernetes/test_pod_generator.py [hunk 5]
```diff
@@ -226,238 +186,169 @@ def test_gen_pod_extract_xcom(self, mock_uuid):
...
-                }],
-                'hostNetwork': False,
-                'imagePullSecrets': [],
-                'volumes': [{
-                    'hostPath': {'path': '/tmp/'},
-                    'name': 'example-kubernetes-test-volume'
-                }],
-            }
-        }, result)
+    def test_from_obj(self):
```
Comment (dimberman) [inline tests/kubernetes/test_pod_generator.py:236]: @kaxil @ashb @potiuk @turbaszek one remaining question is what do we do about the executor_config? Do we force users to inject a k8s.V1Pod object for their executor config and just manually delete a few fields (like pod names)? Or do we continue to allow this dictionary.

File: airflow/example_dags/example_kubernetes_executor_config.py [hunk 6]
```diff
@@ -52,60 +53,67 @@ def test_volume_mount():
     start_task = PythonOperator(
         task_id="start_task",
         python_callable=print_stuff,
-        executor_config={
-            "KubernetesExecutor": {
```
Comment (kaxil) [inline airflow/example_dags/example_kubernetes_executor_config.py:71]: I think having the executor name would be helpful if in future we want to see configs for multiple executors

File: airflow/executors/kubernetes_executor.py [hunk 7]
```diff
@@ -70,116 +69,25 @@ class KubeConfig:  # pylint: disable=too-many-instance-attributes
...
-        self.logs_volume_host = conf.get(self.kubernetes_section, 'logs_volume_host')
-
-        # This prop may optionally be set for PV Claims and is used to write logs
-        self.base_log_folder = conf.get(self.logging_section, 'base_log_folder')
+        self.worker_container_repository = conf.get(
+            self.kubernetes_section, 'worker_container_repository')
+        self.worker_container_tag = conf.get(
+            self.kubernetes_section, 'worker_container_tag')
+        self.kube_image = '{}:{}'.format(
+            self.worker_container_repository, self.worker_container_tag)
```
Comment (kaxil) [inline airflow/executors/kubernetes_executor.py]:
```suggestion
        self.kube_image = f'{self.worker_container_repository}:{self.worker_container_tag}'
```

**Review by kaxil — COMMENTED** [review]
Can you rebase on latest Master please if you haven't already?

File: airflow/providers/cncf/kubernetes/operators/kubernetes_pod.py [hunk 8]
```diff
@@ -365,43 +351,53 @@ def create_new_pod_for_operator(self, labels, launcher) -> Tuple[State, k8s.V1Po
...
+                    restart_policy='Never',
+                    priority_class_name=self.priority_class_name,
+                    volumes=self.volumes,
+                )
+            )
+        for secret in self.secrets:
+            pod = secret.attach_to_pod(pod)
+        if self.do_xcom_push:
+            from airflow.kubernetes.pod_generator import PodGenerator
+            pod = PodGenerator.add_xcom_sidecar(pod)
```
Comment (kaxil) [inline airflow/providers/cncf/kubernetes/operators/kubernetes_pod.py:397]:
```suggestion
            pod = pod_generator.PodGenerator.add_xcom_sidecar(pod)
```

File: telepresence.log [hunk 9]
```diff
@@ -0,0 +1,36 @@
+   0.0 TEL | Telepresence 0.105 launched at Mon Aug 24 20:51:33 2020
```
Comment (kaxil) [inline telepresence.log]: Don't think you wanted to add this file :)

File: airflow/example_dags/example_kubernetes_executor_config.py (hunk 6 above)
Comment (dimberman) [reply airflow/example_dags/example_kubernetes_executor_config.py:71]: The problem I'm seeing is that k8s.V1Pod objects aren't serializable into dictionaries. Not sure how to get around that and would rather not optimize for a future I don't think will come

Comment (dimberman) [reply airflow/example_dags/example_kubernetes_executor_config.py:71]: (at this point the executor_config has existed for years and no other executor has used it. If it wasn't such a pain I'd push for us to change the name to something like "kubernetes_modifier_pod")

Comment (kaxil) [reply airflow/example_dags/example_kubernetes_executor_config.py:71]: > The problem I'm seeing is that k8s.V1Pod objects aren't serializable into dictionaries. Not sure how to get around that and would rather not optimize for a future I don't think will come

We will need it for DAG Serialization too, is it not possible to convert V1Pod to dict ? I thought each k8s object had some kind of method for that

Comment (dimberman) [reply airflow/example_dags/example_kubernetes_executor_config.py:71]: Let me try again. Maybe I'm missing something.

Comment (dimberman) [reply airflow/example_dags/example_kubernetes_executor_config.py:71]:
```  
File "/home/airflow/.local/lib/python3.7/site-packages/airflow/utils/json.py", line 54, in _default
    raise TypeError(f"Object of type '{obj.__class__.__name__}' is not JSON serializable")
TypeError: Object of type 'V1Pod' is not JSON serializable
```

Yeah so if I try to make the pod a field I end up getting this error when I try to render on the page. Not sure what the best way forward is there. Maybe add a field? I don't think we need to change the DB as the executor_config isn't stored in the DB anyways.

Comment (dimberman) [reply airflow/example_dags/example_kubernetes_executor_config.py:71]: (though the downside of adding a field is that it would create even more confusing in transitioning DAGs. I'd personally either figure out a solution to this json serialization issue, or just have multiple types for the executor_config)

Comment (kaxil) [reply airflow/example_dags/example_kubernetes_executor_config.py:71]: We do store executor_config in TaskInstance table:

https://github.com/apache/airflow/blob/d7602654526fdd2876466371404784bd17cfe0d2/airflow/models/taskinstance.py#L222

Although we use `dill` to store it

Comment (kaxil) [reply airflow/example_dags/example_kubernetes_executor_config.py:71]: `k8s.V1Pod(...).to_dict` should work for serialization, isn't it?

https://github.com/kubernetes-client/python/blob/v11.0.0/kubernetes/client/models/v1_pod.py#L179-L201

Comment (dimberman) [reply airflow/example_dags/example_kubernetes_executor_config.py:71]: @kaxil yeah but forcing using to use the to_dict function seems kinda clunky, no?

We store the executor_config in the DB but TBH we don't really need to. Maybe we can just create new fields "kubernetes_executor_config" and "dag_serialization_config" and just remove the "executor_config" as a whole?

Comment (kaxil) [reply airflow/example_dags/example_kubernetes_executor_config.py:71]: No, I meant, adding code in serialization module, https://github.com/apache/airflow/blob/e565368f2e988a06b0398b77e66859c138905ae7/airflow/serialization/serialized_objects.py#L159-L216

that if the object is k8s.V1Pod, store it as `k8s.V1Pod.to_dict()` .

Comment (kaxil) [reply airflow/example_dags/example_kubernetes_executor_config.py:71]: >We store the executor_config in the DB but TBH we don't really need to. Maybe we can just create new fields "kubernetes_executor_config" and "dag_serialization_config" and just remove the "executor_config" as a whole?

If we don't need to store it in DB then sure we can remove it

Comment (dimberman) [reply airflow/example_dags/example_kubernetes_executor_config.py:71]: Ok I got it working with the traditional executor_config. Thanks @kaxil!

File: airflow/kubernetes/pod_generator_deprecated.py [hunk 10]
```diff
@@ -0,0 +1,620 @@
+# Licensed to the Apache Software Foundation (ASF) under one
```
Comment (dimberman) [inline airflow/kubernetes/pod_generator_deprecated.py:1]: @ashb @potiuk @kaxil @turbaszek With this class we can actually maintain the existing executor_config with a deprecation warning until 2.1 or 2.2. This should help ease the transition. WDYT?

**Comment by dimberman** [comment]
@potiuk @kaxil @ashb @turbaszek all tests passing and added documentation. PTAL!

File: UPDATING.md [hunk 11]
```diff
@@ -153,6 +152,175 @@ The Old and New provider configuration keys that have changed are as follows
...
+            ]
+        )
+    )
+    }
+)
+```
+For Airflow 2.0, the traditional `executor_config` will continue operation with a deprecation warning,
+but will be removed in a future version.
+
+### Changes to the KubernetesPodOperator
```
Comment (mik-laj) [inline UPDATING.md:280]: In my opinion, this tutorial will not allow easy migration between the old and the new representation. Many parameters are not described and the new representation will not always be obvious.  The user will have to read the version 1.10 code to understand what they should do now.

File: UPDATING.md [hunk 12]
```diff
@@ -153,6 +152,175 @@ The Old and New provider configuration keys that have changed are as follows
...
+    volumes=[volume],
+    volume_mounts=[volume_mount],
+    name="airflow-test-pod",
+    task_id="task",
+    is_delete_operator_pod=True,
+    hostnetwork=False)
+```
+We decided to keep the Secret class as users seem to really like that simplifies the complexity of mounting
+Kubernetes secrets into workers.
+
```
Comment (mik-laj) [inline UPDATING.md:371]: Can you add a migration guide for each configuration option?

File: airflow/providers/cncf/kubernetes/operators/kubernetes_pod.py [hunk 13]
```diff
@@ -65,11 +60,11 @@ class KubernetesPodOperator(BaseOperator):  # pylint: disable=too-many-instance-
         The docker image's CMD is used if this is not provided.
     :type arguments: list[str]
     :param ports: ports for launched pod.
-    :type ports: list[airflow.kubernetes.pod.Port]
+    :type ports: list[k8s.V1ContainerPort]
```
Comment (mik-laj) [inline airflow/providers/cncf/kubernetes/operators/kubernetes_pod.py:64]: I would be happy if there was an entry with an example for each parameter that has changed. We cannot assume that the user will know these two classes and will be able to migrate on their own.

File: UPDATING.md (hunk 11 above)
Comment (dimberman) [reply UPDATING.md:280]: @mik-laj Thank you for that feedback. I've added more detailed docs on the changed parameters. Can you please take a look when you have some time?

File: UPDATING.md (hunk 12 above)
Comment (dimberman) [reply UPDATING.md:371]: Added!

**Comment by potiuk** [comment]
Hey @dimberman - > should we wait with releasing backports before this one is merged ?

**Comment by dimberman** [comment]
@potiuk yes This is breaking for the KubernetesPodOperator so we should hold off. I'm hoping to have this merged in the next day or two. Can I please get a lookover? :)

**Comment by potiuk** [comment]
On vacations now. But yes. I will be on a lookout :).

File: airflow/utils/json.py [hunk 14]
```diff
@@ -20,6 +20,7 @@
 from datetime import date, datetime
 
 import numpy as np
+from kubernetes.client import models as k8s
```
Comment (ashb) [inline airflow/utils/json.py]:
```suggestion
try:
    from kubernetes.client import models as k8s
except ImportError:
    k8s = None
```

File: airflow/utils/json.py [hunk 15]
```diff
@@ -50,5 +51,8 @@ def _default(obj):
         elif isinstance(obj, (np.float_, np.float16, np.float32, np.float64,
                               np.complex_, np.complex64, np.complex128)):
             return float(obj)
+        elif isinstance(obj, k8s.V1Pod):
```
Comment (ashb) [inline airflow/utils/json.py]:
```suggestion
        elif k8s is not None and isinstance(obj, k8s.V1Pod):
```

File: airflow/serialization/enums.py [hunk 16]
```diff
@@ -42,3 +42,4 @@ class DagAttributeTypes(str, Enum):
     DICT = 'dict'
     SET = 'set'
     TUPLE = 'tuple'
+    POD = 'pod'
```
Comment (ashb) [inline airflow/serialization/enums.py]:
```suggestion
    POD = 'k8s.V1Pod'
```

Maybe?

File: tests/kubernetes/test_pod_generator.py [hunk 17]
```diff
@@ -241,238 +213,224 @@ def test_gen_pod_extract_xcom(self, mock_uuid):
...
                 }],
-                'hostNetwork': False,
-                'imagePullSecrets': [],
                 'volumes': [{
                     'hostPath': {'path': '/tmp/'},
                     'name': 'example-kubernetes-test-volume'
                 }],
             }
         }, result)
+        # TODO: Should we save this feature?
```
Comment (mik-laj) [inline tests/kubernetes/test_pod_generator.py]: ?

File: tests/kubernetes/test_pod_generator.py [hunk 18]
```diff
@@ -241,238 +213,224 @@ def test_gen_pod_extract_xcom(self, mock_uuid):
...
-        base_pod.metadata.name = name
-
-        result = PodGenerator.reconcile_pods(base_pod, mutator_pod)
-        self.assertEqual(base_pod, result)
-
-        mutator_pod = k8s.V1Pod()
-        result = PodGenerator.reconcile_pods(base_pod, mutator_pod)
-        self.assertEqual(base_pod, result)
+        pass
+        # mock_uuid.return_value = self.static_uuid
```
Comment (mik-laj) [inline tests/kubernetes/test_pod_generator.py]: ?

**Comment by dimberman** [comment]
@mik-laj fixed those tests, PTAL :)

File: airflow/example_dags/example_kubernetes_executor_config.py [hunk 19]
```diff
@@ -52,59 +53,68 @@ def test_volume_mount():
...
-                "annotations": {"test": "annotation"}
-            }
+        executor_config={"podOverride": k8s.V1Pod(
```
Comment (kaxil) [inline airflow/example_dags/example_kubernetes_executor_config.py]: After https://github.com/apache/airflow/pull/10756 is merged, we need to change `podOverride` -> `pod_override` in this PR

File: airflow/example_dags/example_kubernetes_executor_config.py [hunk 20]
```diff
@@ -52,59 +53,68 @@ def test_volume_mount():
...
         python_callable=print_stuff,
-        executor_config={
-            "KubernetesExecutor": {
-                "namespace": "test-namespace",
-                "labels": {
-                    "release": "stable"
-                }
-            }
-        }
+        # executor_config={"podOverride": k8s.V1Pod(
```
Comment (kaxil) [inline airflow/example_dags/example_kubernetes_executor_config.py]: is the intention to remove these comments?

File: airflow/executors/kubernetes_executor.py [hunk 21]
```diff
@@ -483,17 +326,23 @@ def run_next(self, next_job: KubernetesJobType) -> None:
...
             dag_id=dag_id,
             task_id=task_id,
+            kube_image=self.kube_config.kube_image,
             try_number=try_number,
             date=execution_date,
             command=command,
-            kube_executor_config=kube_executor_config,
-            worker_config=self.worker_configuration_pod
+            kube_mutation_object=kube_executor_config,
+            base_worker_pod=base_worker_pod
```
Comment (kaxil) [inline airflow/executors/kubernetes_executor.py]: wherever this signature was changed, if we haven't already have it in Updating.md we should add it

Comment (dimberman) [reply airflow/executors/kubernetes_executor.py]: @kaxil do we need to note this if it's not an external class?

File: kubernetes_tests/test_kubernetes_executor.yaml [hunk 22]
```diff
@@ -0,0 +1,82 @@
...
+          readOnly: true
+          subPath: airflow_local_settings.py
+  dnsPolicy: ClusterFirst
+  enableServiceLinks: true
+  imagePullSecrets:
+    - name: airflow-registry
+  restartPolicy: Never
+  securityContext:
+    fsGroup: 50000
+    runAsUser: 50000
```
Comment (ashb) [inline kubernetes_tests/test_kubernetes_executor.yaml]:
```suggestion
    runAsUser: '{{ .Values.uid }}'
    fsGroup: '{{ .Values.gid }}' 
```

etc.

File: chart/values.yaml [hunk 23]
```diff
@@ -541,32 +543,11 @@ config:
...
-    git_sync_depth: '{{ .Values.dags.gitSync.depth }}'
-    git_sync_root: '{{ .Values.dags.gitSync.root }}'
-    git_sync_dest: '{{ .Values.dags.gitSync.dest }}'
-    git_sync_container_repository: '{{ .Values.dags.gitSync.containerRepository }}'
-    git_sync_container_tag: '{{ .Values.dags.gitSync.containerTag }}'
-    git_sync_init_container_name: '{{ .Values.dags.gitSync.containerName }}'
-    git_sync_run_as_user: '{{ .Values.uid }}'
-    git_ssh_known_hosts_configmap_name: '{{- if .Values.dags.gitSync.knownHosts }}{{ include "airflow_config" . }}{{ end }}'
-    git_ssh_key_secret_name: '{{- if .Values.dags.gitSync.sshKeySecret }}{{ .Values.dags.gitSync.sshKeySecret }}{{ end }}'
-    git_sync_credentials_secret: '{{- if .Values.dags.gitSync.credentialsSecret }}{{ .Values.dags.gitSync.credentialsSecret }}{{ end }}'
```
Comment (ashb) [inline chart/values.yaml]: Didn't we discuss that the chart would take these helm values and generate the kube template file based on these?

Comment (dimberman) [reply chart/values.yaml]: @ashb OOF good catch thank you

Comment (dimberman) [reply chart/values.yaml]: @ashb every value in this section is brought from other parts of the helm file. Should we keep this around or delete this section and still have a default pod_template file from those other values (seeing as this section is specifically made for the airflow.cfg)?

Comment (ashb) [reply chart/values.yaml]: This section can go, yeah -- just using this as a place to comment on.

File: airflow/kubernetes/pod_generator.py [hunk 24]
```diff
@@ -123,100 +124,18 @@ class PodGenerator:
...
-        configmaps: Optional[List[str]] = None,
-        dnspolicy: Optional[str] = None,
-        schedulername: Optional[str] = None,
         pod: Optional[k8s.V1Pod] = None,
         pod_template_file: Optional[str] = None,
-        extract_xcom: bool = False,
-        priority_class_name: Optional[str] = None,
+        extract_xcom: bool = True
     ):
         self.validate_pod_generator_args(locals())
```
Comment (davlum) [inline airflow/kubernetes/pod_generator.py]: `validate_pod_generator_args` can be significantly simplified. The only reason it was so complex was because of the signature inspection it was doing for the ridiculous number of arguments being passed. Now the validation is quite simple, the client either passes `pod` or `pod_template_file`.

**Review by davlum — COMMENTED** [review]
If possible I would further simplify this by removing all of the `reconcile_*` methods as they can have unexpected results for the client. The attributes Airflow wants to add at runtime (what's currently called the dynamic pod) do not actually constitute a full pod definition. They are limited and their interactions fairly well understood. The possibilities on configuration now shouldn't have to merge arbitrary pods, which can be troublesome.

The options are as I understand it are;
1. The client configures the pod via `airflow.cfg`, and doesn't pass `executor_config`. Just add the additional dynamic arguments to the pod.
2. The client configures a pod via `airflow.cfg` but does pass a pod via `executor_config`. Ignore the `airflow.cfg` pod, use the `executor_config` pod instead and add the dynamic arguments to that pod.

**Comment by dimberman** [comment]
@davlum interesting, though without the reconcile methods how would users use the pod_override setting, which is an arbitrary pod? We need some way to ensure features are either added or overwritten based on these three stages. It's also worth mentioning that we now have a `generate_dag_yaml` command in the CLI so users can see what their pods will look like before they launch (and we plan to add this preview to the UI)

**Comment by davlum** [comment]
Pod override would do exactly that, it would completely override the base definition, not merge with it.

**Comment by dimberman** [comment]
@davlum I don't think we would want the pod_override to override all values. The examples I can think of off the top of my head are labels and volume_mounts. The base pod file will have volume mounts for the airflow.cfg and other potential secrets. I as a non-admin should be allowed to ADD volume mounts, but I shouldn't be able to delete them via override. Same with labels. Admin might have labels or node affinities they don't want messed with.

While yes it's POSSIBLE to ensure those are there using the pod_mutation_hook, doing so places onus on the admin for the user's messups.

File: UPDATING.md [hunk 25]
```diff
@@ -154,6 +153,480 @@ The Old and New provider configuration keys that have changed are as follows
...
 For more information, visit https://flask-appbuilder.readthedocs.io/en/latest/security.html#authentication-oauth
 
+### Changes to the KubernetesExecutor
+
+#### The KubernetesExecutor Will No Longer Read from the airflow.cfg for Base Pod Configurations
+
+In Airflow 2.0, the KubernetesExecutor will require a base pod template written in yaml. This file can exist
+anywhere on the host machine and will be linked using the `pod_template_file` configuration in the airflow.cfg.
+
+The airflow.cfg will still accept values for the `worker_container_repository`, the `worker_container_tag`, and
```
Comment (kaxil) [inline UPDATING.md]:
```suggestion
The `airflow.cfg` will still accept values for the `worker_container_repository`, the `worker_container_tag`, and
```

File: UPDATING.md [hunk 26]
```diff
@@ -154,6 +153,480 @@ The Old and New provider configuration keys that have changed are as follows
...
+the default namespace.
+
+#### The executor_config Will Now Expect a `kubernetes.client.models.V1Pod` Class When Launching Tasks
```
Comment (kaxil) [inline UPDATING.md]:
```suggestion
#### The `executor_config` Will Now Expect a `kubernetes.client.models.V1Pod` Class When Launching Tasks
```

File: UPDATING.md [hunk 27]
```diff
@@ -154,6 +153,480 @@ The Old and New provider configuration keys that have changed are as follows
...
+
+In airflow 1.10, users could modify task pods at runtime by passing a dictionary to the `executor_config` variable.
```
Comment (kaxil) [inline UPDATING.md]:
```suggestion
In Airflow 1.10.x, users could modify task pods at runtime by passing a dictionary to the `executor_config` variable.
```

File: UPDATING.md [hunk 28]
```diff
@@ -154,6 +153,480 @@ The Old and New provider configuration keys that have changed are as follows
...
+    )
+    }
+)
+```
+For Airflow 2.0, the traditional `executor_config` will continue operation with a deprecation warning,
+but will be removed in a future version.
+
+### Changes to the KubernetesPodOperator
+
+Much like the KubernetesExecutor, the KubernetesPodOperator will no longer take Airflow custom classes and will
```
Comment (kaxil) [inline UPDATING.md]:
```suggestion
Much like the `KubernetesExecutor`, the `KubernetesPodOperator` will no longer take Airflow custom classes and will
```

File: airflow/kubernetes/pod_generator.py (hunk 24 above)
Comment (dimberman) [reply airflow/kubernetes/pod_generator.py]: @davlum good point, fixed!

File: UPDATING.md [hunk 29]
```diff
@@ -154,6 +153,480 @@ The Old and New provider configuration keys that have changed are as follows
 
 For more information, visit https://flask-appbuilder.readthedocs.io/en/latest/security.html#authentication-oauth
 
+### Changes to the KubernetesExecutor
+
+#### The KubernetesExecutor Will No Longer Read from the airflow.cfg for Base Pod Configurations
```
Comment (mik-laj) [inline UPDATING.md:158]: Can you add a full list of configuration options that have been deleted? This is important if we are going to work on `airflow upgrade-check`, because having the list we can detect incorrect configuration and notify the user at runtime.

File: UPDATING.md [hunk 30]
```diff
@@ -154,6 +153,480 @@ The Old and New provider configuration keys that have changed are as follows
...
+                        path="/tmp/"
+                    )
+                )
+            ]
+        )
+    )
+    }
+)
+```
+For Airflow 2.0, the traditional `executor_config` will continue operation with a deprecation warning,
```
Comment (mik-laj) [inline UPDATING.md:277]: Do we have this deprecation warning? I didn't see it in the code, but I might have missed it.

File: docs/howto/operator/kubernetes.rst [hunk 31]
```diff
@@ -53,13 +53,10 @@ Ultimately, it allows Airflow to act a job orchestrator - no matter the language
 
 How to use cluster ConfigMaps, Secrets, and Volumes with Pod?
 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
-Kubernetes cluster resources such as ConfigMaps, Secrets, and Volumes can be used with a Pod to be launched.
+Kubernetes cluster resources such as Secrets can be used with a Pod to be launched.
 Utilize the Airflow Kubernetes model classes such as:
```
Comment (mik-laj) [inline docs/howto/operator/kubernetes.rst]: This sentence needs to be corrected. The plural is used when one class is described.

Comment (ashb) [reply docs/howto/operator/kubernetes.rst]: Do we still support standard python dicts? (Line 60)

Also: why did this bit need to change -- should `Volume`  or `VolumeMount` not be used anymore?

File: UPDATING.md (hunk 30 above)
Comment (dimberman) [reply UPDATING.md:277]: Hi @mik-laj, the deprecation warning was already merged so it didn't show up in the PR https://github.com/apache/airflow/pull/10393/files#diff-868ed785b2a336f20cb6a577dde2502aR198-R201

File: docs/howto/operator/kubernetes.rst (hunk 31 above)
Comment (dimberman) [reply docs/howto/operator/kubernetes.rst]: Fixed. @ashb yes `Volume` and `VolumeMount` would now be `k8s.V1Volume` and `k8s.V1VolumeMount`, I go into more detail on this in the UPDATING.md

File: UPDATING.md (hunk 29 above)
Comment (dimberman) [reply UPDATING.md:158]: @mik-laj added

**Comment by dimberman** [comment]
@mik-laj can I get another lookover? :)

File: airflow/kubernetes/pod_generator.py [hunk 32]
```diff
@@ -396,7 +240,7 @@ def from_legacy_obj(obj) -> Optional[k8s.V1Pod]:
                     limits=limits
                 )
         namespaced['resources'] = resources
-        return PodGenerator(**namespaced).gen_pod()
+        return PodGeneratorDeprecated(**namespaced).gen_pod()
```
Comment (ashb) [inline airflow/kubernetes/pod_generator.py:243]: Can/should we issue a FutureDeprecationWarning here, so that we can remove this code path in 2.1/2.2?

Comment (dimberman) [reply airflow/kubernetes/pod_generator.py:243]: @ashb there is a future deprecation https://github.com/apache/airflow/pull/10393/files#diff-868ed785b2a336f20cb6a577dde2502aR198-R201

Comment (ashb) [reply airflow/kubernetes/pod_generator.py:243]: Oh, you issue a warning a few lines up (from the previous PR. right?)

Comment (dimberman) [reply airflow/kubernetes/pod_generator.py:243]: Yep :)

**Review by ashb — APPROVED** [review]

**Review by kaxil — APPROVED** [review]
